packages/kolme/src/gossip/sync_manager.rs

```diff
@@ -0,0 +1,623 @@
+        if let Some(WaitingBlock::Needed(status)) = self.needed_blocks.get_mut(&height) {
+            status.remove_peer(peer);
+            self.trigger.trigger();
+        }
+    }
+
+    pub(super) async fn add_pending_block(
+        &mut self,
+        gossip: &Gossip<App>,
+        block: Arc<SignedBlock<App::Message>>,
+        peer: Option<PeerId>,
+    ) {
+        let height = block.height();
+
+        match gossip.kolme.has_block(height).await {
+            Err(e) => {
+                tracing::warn!(
+                    "{}: add_pending_block: has_block({height}) failed: {e}",
+                    gossip.local_display_name
+                );
+                return;
+            }
+            Ok(true) => {
+                self.needed_blocks.remove(&height);
+                return;
+            }
+            Ok(false) => (),
+        }
+
+        let Some(waiting) = self.needed_blocks.get_mut(&height) else {
+            return;
+        };
+
+        match waiting {
+            // We needed this block, so let's do it.
+            WaitingBlock::Needed(_) => {
+                *waiting = WaitingBlock::Received { block, peer };
+                self.trigger.trigger();
+            }
+            // We already got it, nothing to do.
+            WaitingBlock::Received { .. } | WaitingBlock::Pending(_) => (),
+        }
+    }
+
+    pub(super) async fn add_merkle_layer(
+        &mut self,
+        gossip: &Gossip<App>,
+        hash: Sha256Hash,
+        layer: MerkleLayerContents,
+        peer: PeerId,
+    ) -> Result<()> {
+        let Some(mut entry) = self.needed_blocks.first_entry() else {
+            return Ok(());
+        };
+
+        let pending = match entry.get_mut() {
+            WaitingBlock::Needed(_) | WaitingBlock::Received { .. } => return Ok(()),
+            WaitingBlock::Pending(pending) => pending,
+        };
+
+        if !pending.needed_layers.contains_key(&hash) || pending.pending_layers.contains_key(&hash)
+        {
+            return Ok(());
+        }
+
+        if !gossip.kolme.has_merkle_hash(hash).await? {
+            let mut has_all = true;
+            for child in &layer.children {
+                if !gossip.kolme.has_merkle_hash(*child).await? {
+                    has_all = false;
+                    pending
+                        .reverse_layers
+                        .entry(*child)
+                        .or_default()
+                        .insert(hash);
+                    pending
+                        .needed_layers
+                        .entry(*child)
+                        .or_insert(RequestStatus::new(Some(peer)));
+                }
+            }
+
+            pending.pending_layers.insert(hash, layer);
+            pending.needed_layers.remove(&hash);
+
+            if has_all {
+                Self::process_available_hash(gossip, pending, hash).await?;
+                self.trigger.trigger();
```

Comment:
Probably this trigger can be skipped since in line number 307 we are anyway going to trigger ?